Add cluster mode support

qjsoq opened the pull request:
This pull request introduces a new, comprehensive Helm chart named valkey-cluster for deploying a clustered Valkey instance on Kubernetes.
This chart is designed to handle the complexities of a clustered setup by leveraging a StatefulSet and an intelligent initialization script. 

What This PR Does:

- **StatefulSet Deployment**: Ensures stable network identifiers and ordered deployment, which is critical for a clustered database.
- Introduces **Dynamic Per-Pod PVCs**: Implements volumeClaimTemplates to automatically provision a unique PersistentVolumeClaim for each cluster node, ensuring data persistence and isolation.
- **Automated Cluster Initialization**: A new init-cluster.sh script (managed via cluster_config.yaml) is introduced. This script handles complex cluster bootstrapping, node discovery, and dynamic joining logic for new or replacement pods.
- **Graceful Master Failover**: Adds a default preStop lifecycle hook. This critical feature ensures that when a master pod is scheduled for termination (e.g., during a rolling update), it first triggers a controlled failover to one of its slaves before shutting down, minimizing downtime.
- **Adds Headless Service**: Creates the necessary headless service for stable network discovery between pods within the StatefulSet.
- **Exposes Metrics Exporter**: Adds a dedicated Kubernetes Service for the Prometheus metrics exporter sidecar. This makes cluster metrics easily discoverable and scrape-able for monitoring systems.
- Adds a **PodDisruptionBudget** to protect the cluster during voluntary disruptions.

**How the Auto-Clustering Works**
The core of the cluster logic resides in the` init-cluster.sh` script, which runs as a background process in the main container:

Peer Discovery: Each pod uses the Kubernetes headless service to discover its peers.

Joining an Existing Cluster: If a new pod starts and discovers an already healthy cluster, it uses the `CLUSTER MEET` command to join. It then intelligently finds a master with a deficit of replicas and assigns itself as a replica using `CLUSTER REPLICATE`.

Initial Cluster Creation: If no cluster is found, the first pod (<release-name>-0) takes on the role of the initiator. It waits for all other pods to become ready and then bootstraps the entire cluster using the `valkey-cli --cluster create` command with the appropriate number of replicas.

Resiliency: The script can handle pods restarting by attempting to re-join the cluster and forget its old failed entry if necessary.

**Here are some screenshots demonstrating the chart in action after a successful deployment.**
For deployment these settings were set:
```
replicaCount: 6
metrics:
  enabled: true
dataStorage:
  enabled: true
  requestedSize: "8Gi"
```

1. Cluster Nodes Output
<img width="825" height="144" alt="image" src="https://github.com/user-attachments/assets/0580b318-ad9a-4cf8-b1d4-fe4b49827a87" />

This shows the successful formation of a 6-node cluster (3 masters, 3 replicas).

2. Running Pods and PVCs
<img width="890" height="265" alt="image" src="https://github.com/user-attachments/assets/af31b9f2-19e7-4447-83ed-7b1b6136ebf3" />
<img width="792" height="273" alt="image" src="https://github.com/user-attachments/assets/c47c78f3-963a-4739-a721-b106abff8ee4" />

This confirms that the StatefulSet has successfully created all pods and their corresponding PersistentVolumeClaims.

3. Services
<img width="815" height="77" alt="image" src="https://github.com/user-attachments/assets/f0f637c9-e5a9-4a97-a26e-7880f1e5ba99" />

qjsoq committed:
Add cluster support

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Enhance deployment type management

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Enhance functions namings

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Fix number of replicas

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Fix Quick Start test

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Change default replicaCount

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Update README

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Fix namespace inconsistency

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Update main readme

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Fix connectivity issue

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Adjust Valkey configuration

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Adjust cluster creation logic

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Fix minor cluster issues

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Add base image for valkey

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Add readiness probe

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Move script files and add scale up functionality

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

lyatanski commented:
Wouldn't it be slightly more optimal if there is `podManagementPolicy: Parallel` set in the `spec`? On initial installation this should improve cluster creation time.

qjsoq committed:
Merge remote-tracking branch 'upstream/main' into cluster-enable-branch

qjsoq committed:
Merge branch 'main' into cluster-enable-branch

qjsoq committed:
Add podManagementPolicy option

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Remove default image tag from values

Signed-off-by: Dmytro Artamonov <artamonovdima355@gmail.com>

qjsoq committed:
Add basic test

qjsoq committed:
Merge remote-tracking branch 'upstream/main' into cluster-enable-branch

qjsoq committed:
Merge remote-tracking branch 'upstream/main' into cluster-enable-branch

volodymyr-babak committed:
fix(cluster): survive simultaneous full restart with persistent storage

When all pods restart at once (chaotic delete / full node loss) their IPs
change together, but nodes.conf records peers at stale IPs, so the cluster
bus cannot reconnect and cluster_state stays 'fail'. Add:
- cluster-announce-hostname (stable headless FQDN) + preferred-endpoint-type
- publishNotReadyAddresses on the headless Service (resolve peers at startup)
- init-container nodes.conf IP fix-up: re-resolve each peer hostname to its
  current IP before valkey-server starts, so the bus reconnects immediately

Validated on QS/Test: with PVC + podManagementPolicy=Parallel, two consecutive
force-deletes of all 6 pods fully self-heal (cluster_state:ok, identical node IDs).

volodymyr-babak committed:
refactor(cluster): make stable-addressing toggleable; disable hostname client endpoint

Reliability testing (8 chaotic full-restart cycles per config) on QS/Test showed:
- nodes.conf IP fix-up is ESSENTIAL (0/3 recover without it)
- cluster-preferred-endpoint-type hostname HURTS recovery (1/3 with it, 8/8 without)
  — forcing the cluster bus onto hostnames adds DNS churn during a simultaneous
  restart; correcting nodes.conf IPs (bus stays on IP) is reliable.

Minimum reliable set for simultaneous-restart self-heal:
PVC + podManagementPolicy=Parallel + cluster-announce-hostname +
publishNotReadyAddresses + nodes.conf IP fix-up. Each piece is now an independent
clusterStableAddressing.* toggle; preferHostnameEndpoint defaults OFF.

qjsoq committed:
Merge pull request #2 from volodymyr-babak/fix/valkey-simultaneous-restart-resilience

fix(cluster): survive simultaneous full restart with stable addressing

qjsoq committed:
Merge remote-tracking branch 'origin/cluster-enable-branch' into cluster-enable-branch

qjsoq committed:
Correct topologySpreadConstraints schema, POSIX init.sh, and pod annotations; add unit tests

qjsoq committed:
Merge remote-tracking branch 'upstream/main' into cluster-enable-branch

# Conflicts:
#	Justfile

qjsoq committed:
Update the chart and app versions
